resourcettl/cache.go

```diff
@@ -0,0 +1,187 @@
+//
+// The cache keeps track of the resources' usage and makes sure that
+// expired resources are not cleaned up while they are still in use
+// and cleaned up only when they are no longer needed (zero checkouts).
+//
+// Resources with any of following methods can be cleaned up:
+//   - Cleanup()
+//   - Close()
+//   - Close() error
+//   - Stop()
+//   - Stop() error
+func NewCache[K comparable, R any](new func(key K) (R, error), ttl time.Duration) *Cache[K, R] {
+	c := &Cache[K, R]{
+		new:            new,
+		keyMu:          kitsync.NewPartitionLocker(),
+		latestInstance: make(map[K]string),
+		resources:      make(map[string]R),
+		checkouts:      make(map[string]int),
+		expiries:       make(map[string]struct{}),
+		ttl:            ttl,
+		ttlcache:       cachettl.New[K, string](),
+	}
+	c.ttlcache.OnEvicted(c.onEvicted)
+	return c
+}
+
+// Cache is a cache for resources that need to be closed/cleaned-up after expiration.
+//
+// The cache keeps track of the resources' usage and makes sure that
+// expired resources are not cleaned up while they are still in use
+// and cleaned up only when they are no longer needed (zero checkouts).
+//
+// Resources with any of following methods can be cleaned up:
+//   - Cleanup()
+//   - Close()
+//   - Close() error
+//   - Stop()
+//   - Stop() error
+type Cache[K comparable, R any] struct {
+	new func(key K) (R, error) // creates a new resource
+
+	// synchronizes access to the cache for a given key. This is to
+	// allow multiple go-routines to access the cache concurrently for different keys, but still
+	// avoid multiple go-routines creating multiple resources for the same key.
+	keyMu *kitsync.PartitionLocker
+
+	mu             sync.RWMutex        // protects the following maps
+	latestInstance map[K]string        // maps a key to its latest instanceID. Multiple instances can co-exist for the same key (due to expired resources which might still be checked out)
+	resources      map[string]R        // maps an instanceID to its resource
+	checkouts      map[string]int      // keeps track of how many checkouts are there for a given instanceID
+	expiries       map[string]struct{} // keeps track of instances that are expired and need to be cleaned up after all checkouts are done
+
+	ttl      time.Duration
+	ttlcache *cachettl.Cache[K, string]
+}
+
+// Checkout returns a resource for the given key. If the resource is not available, it creates a new one.
+// The caller must call the returned checkin function when the resource is no longer needed, to release the resource.
+// Multiple checkouts for the same key are allowed and they can all share the same resource. The resource is cleaned up
+// only when all checkouts are checked-in and the resource's ttl has expired (or its key has been invalidated through [Invalidate]).
+func (c *Cache[K, R]) Checkout(key K) (resource R, checkin func(), err error) {
+	defer c.lockKey(key)()
+
+	if instanceID := c.ttlcache.Get(key); instanceID != "" {
```

Comment:
The name of this variable seems implementation specific while the `resourcettl.Cache` seems generic. Wouldn't it make more sense to rename `instanceID` to something like `resource` or `value`?